File: graph_shortest_path/dijkstra/src/DijkstraDemo.cs
// C#
// グラフの最短経路: ダイクストラ法 (dijkstra)

using System;
using System.Collections.Generic;
using System.Linq;

// グラフの最短経路: ダイクストラ法 (dijkstra)
class GraphData
{
    // 隣接ノードとその辺の重みを格納します。
    private Dictionary<string, List<Tuple<string, int>>> _data;

    public GraphData()
    {
        _data = new Dictionary<string, List<Tuple<string, int>>>();
    }

    public Dictionary<string, List<Tuple<string, int>>> Get()
    {
        // グラフの内部データを取得します。
        return _data;
    }

    public List<string> GetVertices()
    {
        // グラフの全頂点をリストとして返します。
        return _data.Keys.ToList();
    }

    public List<Tuple<string, string, int>> GetEdges()
    {
        // グラフの全辺をリストとして返します。
        // 無向グラフの場合、(u, v, weight) の形式で返します。
        // 重複を避けるためにハッシュセットを使用します。
        var edges = new HashSet<Tuple<string, string, int>>();
        
        foreach (var vertex in _data.Keys)
        {
            foreach (var neighborInfo in _data[vertex])
            {
                var neighbor = neighborInfo.Item1;
                var weight = neighborInfo.Item2;
                
                // 辺を正規化してセットに追加 (小さい方の頂点を最初にするなど)
                string[] sortedVertices = new[] { vertex, neighbor };
                Array.Sort(sortedVertices);
                
                edges.Add(new Tuple<string, string, int>(sortedVertices[0], sortedVertices[1], weight));
            }
        }
        
        return edges.ToList();
    }

    public List<Tuple<string, int>> GetNeighbors(string vertex)
    {
        // 指定された頂点の隣接ノードと辺の重みのリストを返します。
        // 形式: [(隣接頂点, 重み), ...]
        if (_data.ContainsKey(vertex))
        {
            return _data[vertex];
        }
        else
        {
            return null; // 頂点が存在しない場合はNullを返す
        }
    }

    public bool AddVertex(string vertex)
    {
        // 新しい頂点をグラフに追加します。
        if (!_data.ContainsKey(vertex))
        {
            _data[vertex] = new List<Tuple<string, int>>();
            return true;
        }
        
        // 既に存在する場合は追加しないがTrueを返す（変更なしでも成功とみなす）
        return true;
    }

    public bool AddEdge(string vertex1, string vertex2, int weight)
    {
        // 両頂点間に辺を追加します。重みを指定します。
        // 頂点がグラフに存在しない場合は追加します。
        if (!_data.ContainsKey(vertex1))
        {
            AddVertex(vertex1);
        }
        
        if (!_data.ContainsKey(vertex2))
        {
            AddVertex(vertex2);
        }

        // vertex1 -> vertex2 の辺を追加（重み付き）
        bool edgeExistsV1V2 = false;
        for (int i = 0; i < _data[vertex1].Count; i++)
        {
            if (_data[vertex1][i].Item1 == vertex2)
            {
                _data[vertex1][i] = new Tuple<string, int>(vertex2, weight); // 既に存在する場合は重みを更新
                edgeExistsV1V2 = true;
                break;
            }
        }
        
        if (!edgeExistsV1V2)
        {
            _data[vertex1].Add(new Tuple<string, int>(vertex2, weight));
        }

        // vertex2 -> vertex1 の辺を追加（重み付き）
        bool edgeExistsV2V1 = false;
        for (int i = 0; i < _data[vertex2].Count; i++)
        {
            if (_data[vertex2][i].Item1 == vertex1)
            {
                _data[vertex2][i] = new Tuple<string, int>(vertex1, weight); // 既に存在する場合は重みを更新
                edgeExistsV2V1 = true;
                break;
            }
        }
        
        if (!edgeExistsV2V1)
        {
            _data[vertex2].Add(new Tuple<string, int>(vertex1, weight));
        }
        
        return true;
    }

    public bool Clear()
    {
        // グラフを空にします。
        _data.Clear();
        return true;
    }

    public Tuple<List<string>, int> GetShortestPath(string startVertex, string endVertex, Func<string, string, int> heuristic)
    {
        if (!_data.ContainsKey(startVertex) || !_data.ContainsKey(endVertex))
        {
            Console.WriteLine($"ERROR: 開始頂点 '{startVertex}' または 終了頂点 '{endVertex}' がグラフに存在しません。");
            return new Tuple<List<string>, int>(null, int.MaxValue);
        }

        // 距離を初期化: 全ての頂点への距離を無限大に設定し、開始頂点のみ0とする
        var distances = new Dictionary<string, int>();
        foreach (var vertex in _data.Keys)
        {
            distances[vertex] = int.MaxValue;
        }
        distances[startVertex] = 0;

        // 経路を記録するための辞書: 各頂点に至る直前の頂点を保持
        var predecessors = new Dictionary<string, string>();
        foreach (var vertex in _data.Keys)
        {
            predecessors[vertex] = null;
        }

        // 優先度付きキュー: C#のPriorityQueueを使用
        // .NET 6以上であればPriorityQueue<TElement, TPriority>クラスを使用できる
        // ここでは代替としてSortedSetを使う
        var priorityQueue = new SortedSet<Tuple<int, string>>(
            Comparer<Tuple<int, string>>.Create((a, b) => 
            {
                int distanceComparison = a.Item1.CompareTo(b.Item1);
                // 距離が同じ場合は頂点で比較（重複を避けるため）
                return distanceComparison != 0 ? distanceComparison : a.Item2.CompareTo(b.Item2);
            })
        );
        
        priorityQueue.Add(new Tuple<int, string>(0, startVertex)); // (開始頂点への距離, 開始頂点)

        while (priorityQueue.Count > 0)
        {
            // 優先度付きキューから最も距離の小さい頂点を取り出す
            var current = priorityQueue.Min;
            priorityQueue.Remove(current);
            int currentDistance = current.Item1;
            string currentVertex = current.Item2;

            // 取り出した頂点への距離が、すでに記録されている距離より大きい場合は、
            // より短い経路が既に見つかっているためスキップ
            if (currentDistance > distances[currentVertex])
            {
                continue;
            }

            // 終了頂点に到達したら探索終了
            if (currentVertex == endVertex)
            {
                break; // 最短経路が見つかった
            }

            // 現在の頂点から到達可能な隣接頂点を探索
            var neighbors = GetNeighbors(currentVertex);
            if (neighbors != null)
            {
                foreach (var neighborInfo in neighbors)
                {
                    string neighbor = neighborInfo.Item1;
                    int weight = neighborInfo.Item2;
                    int distanceThroughCurrent = distances[currentVertex] + weight;

                    // より短い経路が見つかった場合
                    if (distanceThroughCurrent < distances[neighbor])
                    {
                        distances[neighbor] = distanceThroughCurrent;
                        predecessors[neighbor] = currentVertex;
                        // 優先度付きキューに隣接頂点を追加または更新
                        // ダイクストラ法では heuristic は使用しない (または h=0)
                        priorityQueue.Add(new Tuple<int, string>(distanceThroughCurrent, neighbor));
                    }
                }
            }
        }

        // 終了頂点への最短距離が無限大のままなら、到達不可能
        if (distances[endVertex] == int.MaxValue)
        {
            Console.WriteLine($"INFO: 開始頂点 '{startVertex}' から 終了頂点 '{endVertex}' への経路は存在しません。");
            return new Tuple<List<string>, int>(null, int.MaxValue);
        }

        // 最短経路を再構築
        var path = new List<string>();
        string current = endVertex;
        while (current != null)
        {
            path.Add(current);
            current = predecessors[current];
        }
        path.Reverse(); // 経路は逆順に構築されたので反転

        // 開始ノードから開始されていることを確認
        if (path[0] != startVertex)
        {
            // これが発生するのは、開始頂点が存在しないか、到達不能な場合だが、
            // 前段のチェックで大部分はカバーされているはず。
            // ここに来る場合は、特殊なケース（例えば孤立した開始点と終了点）が考えられる。
            return new Tuple<List<string>, int>(null, int.MaxValue);
        }

        return new Tuple<List<string>, int>(path, distances[endVertex]);
    }
}

class Program
{
    // ヒューリスティック関数 (この例では常に0、ダイクストラ法と同じ)
    // 実際のA*では、問題に応じた適切な推定関数を使用する必要があります。
    static int DummyHeuristic(string u, string v)
    {
        // u と v の間に何らかの推定距離を計算する関数
        // ここではダミーとして常に0を返す
        return 0;
    }

    static void Main(string[] args)
    {
        Console.WriteLine("Dijkstra -----> start");

        var graphData = new GraphData();

        graphData.Clear();
        var inputList1 = new List<Tuple<string, string, int>>
        {
            new Tuple<string, string, int>("A", "B", 4),
            new Tuple<string, string, int>("B", "C", 3),
            new Tuple<string, string, int>("B", "D", 2),
            new Tuple<string, string, int>("D", "A", 1),
            new Tuple<string, string, int>("A", "C", 2),
            new Tuple<string, string, int>("B", "D", 2)
        };
        
        foreach (var input in inputList1)
        {
            graphData.AddEdge(input.Item1, input.Item2, input.Item3);
        }
        
        Console.WriteLine("\nグラフの頂点: [{0}]", string.Join(", ", graphData.GetVertices()));
        Console.WriteLine("グラフの辺 (重み付き): [{0}]", string.Join(", ", graphData.GetEdges().Select(e => $"({e.Item1}, {e.Item2}, {e.Item3})")));
        
        var input = new Tuple<string, string>("A", "B");
        var shortestPath = graphData.GetShortestPath(input.Item1, input.Item2, DummyHeuristic);
        Console.WriteLine($"経路{input.Item1}-{input.Item2} の最短経路は [{string.Join(", ", shortestPath.Item1 ?? new List<string>())}] (重み: {(shortestPath.Item2 == int.MaxValue ? "inf" : shortestPath.Item2.ToString())})");

        graphData.Clear();
        var inputList2 = new List<Tuple<string, string, int>>
        {
            new Tuple<string, string, int>("A", "B", 4),
            new Tuple<string, string, int>("C", "D", 4),
            new Tuple<string, string, int>("E", "F", 1),
            new Tuple<string, string, int>("F", "G", 1)
        };
        
        foreach (var inp in inputList2)
        {
            graphData.AddEdge(inp.Item1, inp.Item2, inp.Item3);
        }
        
        Console.WriteLine("\nグラフの頂点: [{0}]", string.Join(", ", graphData.GetVertices()));
        Console.WriteLine("グラフの辺 (重み付き): [{0}]", string.Join(", ", graphData.GetEdges().Select(e => $"({e.Item1}, {e.Item2}, {e.Item3})")));
        
        input = new Tuple<string, string>("A", "B");
        shortestPath = graphData.GetShortestPath(input.Item1, input.Item2, DummyHeuristic);
        Console.WriteLine($"経路{input.Item1}-{input.Item2} の最短経路は [{string.Join(", ", shortestPath.Item1 ?? new List<string>())}] (重み: {(shortestPath.Item2 == int.MaxValue ? "inf" : shortestPath.Item2.ToString())})");

        graphData.Clear();
        var inputList3 = new List<Tuple<string, string, int>>
        {
            new Tuple<string, string, int>("A", "B", 4),
            new Tuple<string, string, int>("B", "C", 3),
            new Tuple<string, string, int>("D", "E", 5)
        };
        
        foreach (var inp in inputList3)
        {
            graphData.AddEdge(inp.Item1, inp.Item2, inp.Item3);
        }
        
        Console.WriteLine("\nグラフの頂点: [{0}]", string.Join(", ", graphData.GetVertices()));
        Console.WriteLine("グラフの辺 (重み付き): [{0}]", string.Join(", ", graphData.GetEdges().Select(e => $"({e.Item1}, {e.Item2}, {e.Item3})")));
        
        input = new Tuple<string, string>("A", "D");
        shortestPath = graphData.GetShortestPath(input.Item1, input.Item2, DummyHeuristic);
        Console.WriteLine($"経路{input.Item1}-{input.Item2} の最短経路は [{string.Join(", ", shortestPath.Item1 ?? new List<string>())}] (重み: {(shortestPath.Item2 == int.MaxValue ? "inf" : shortestPath.Item2.ToString())})");

        graphData.Clear();
        // 空のリスト
        
        Console.WriteLine("\nグラフの頂点: [{0}]", string.Join(", ", graphData.GetVertices()));
        Console.WriteLine("グラフの辺 (重み付き): [{0}]", string.Join(", ", graphData.GetEdges().Select(e => $"({e.Item1}, {e.Item2}, {e.Item3})")));
        
        input = new Tuple<string, string>("A", "B");
        shortestPath = graphData.GetShortestPath(input.Item1, input.Item2, DummyHeuristic);
        Console.WriteLine($"経路{input.Item1}-{input.Item2} の最短経路は [{string.Join(", ", shortestPath.Item1 ?? new List<string>())}] (重み: {(shortestPath.Item2 == int.MaxValue ? "inf" : shortestPath.Item2.ToString())})");

        Console.WriteLine("\nDijkstra <----- end");
    }
}

File: graph_shortest_path/dijkstra/src/DijkstraDemo.rs
// Rust
// グラフの最短経路: ダイクストラ法 (dijkstra)

use std::cmp::Ordering;
use std::collections::{BinaryHeap, HashMap};

// BinaryHeapでの優先度付きキューの要素のための構造体
#[derive(Copy, Clone, Eq, PartialEq)]
struct State {
    cost: u32,
    vertex: String,
}

// BinaryHeapは最大ヒープなので、逆順で比較する必要がある
impl Ord for State {
    fn cmp(&self, other: &Self) -> Ordering {
        // コストが小さい方を優先するため、otherと自分を比較
        other.cost.cmp(&self.cost)
            .then_with(|| self.vertex.cmp(&other.vertex))
    }
}

impl PartialOrd for State {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        Some(self.cmp(other))
    }
}

struct GraphData {
    // 隣接ノードとその辺の重みを格納します
    // キーは頂点、値はその頂点に隣接する頂点と重みのベクター
    data: HashMap<String, Vec<(String, u32)>>,
}

impl GraphData {
    fn new() -> Self {
        GraphData {
            data: HashMap::new(),
        }
    }

    fn get(&self) -> &HashMap<String, Vec<(String, u32)>> {
        &self.data
    }

    fn get_vertices(&self) -> Vec<String> {
        self.data.keys().cloned().collect()
    }

    fn get_edges(&self) -> Vec<(String, String, u32)> {
        let mut edges = Vec::new();
        let mut seen = std::collections::HashSet::new();

        for (vertex, neighbors) in &self.data {
            for (neighbor, weight) in neighbors {
                // 辺を正規化して重複を避ける
                let (v1, v2) = if vertex < neighbor {
                    (vertex.clone(), neighbor.clone())
                } else {
                    (neighbor.clone(), vertex.clone())
                };

                let edge_key = format!("{}-{}", v1, v2);
                if !seen.contains(&edge_key) {
                    edges.push((v1, v2, *weight));
                    seen.insert(edge_key);
                }
            }
        }
        edges
    }

    fn get_neighbors(&self, vertex: &str) -> Option<&Vec<(String, u32)>> {
        self.data.get(vertex)
    }

    fn add_vertex(&mut self, vertex: &str) -> bool {
        self.data.entry(vertex.to_string()).or_insert_with(Vec::new);
        true
    }

    fn add_edge(&mut self, vertex1: &str, vertex2: &str, weight: u32) -> bool {
        // 頂点がグラフに存在しない場合は追加
        if !self.data.contains_key(vertex1) {
            self.add_vertex(vertex1);
        }
        if !self.data.contains_key(vertex2) {
            self.add_vertex(vertex2);
        }
        
        // vertex1 -> vertex2 の辺を追加または更新
        let mut edge_exists_v1v2 = false;
        if let Some(neighbors) = self.data.get_mut(vertex1) {
            for i in 0..neighbors.len() {
                if neighbors[i].0 == vertex2 {
                    neighbors[i].1 = weight;
                    edge_exists_v1v2 = true;
                    break;
                }
            }
            if !edge_exists_v1v2 {
                neighbors.push((vertex2.to_string(), weight));
            }
        }

        // vertex2 -> vertex1 の辺を追加または更新
        let mut edge_exists_v2v1 = false;
        if let Some(neighbors) = self.data.get_mut(vertex2) {
            for i in 0..neighbors.len() {
                if neighbors[i].0 == vertex1 {
                    neighbors[i].1 = weight;
                    edge_exists_v2v1 = true;
                    break;
                }
            }
            if !edge_exists_v2v1 {
                neighbors.push((vertex1.to_string(), weight));
            }
        }

        true
    }

    fn clear(&mut self) -> bool {
        self.data.clear();
        true
    }

    fn get_shortest_path(&self, start_vertex: &str, end_vertex: &str, _heuristic: fn(&str, &str) -> u32) -> (Option<Vec<String>>, u32) {
        if !self.data.contains_key(start_vertex) || !self.data.contains_key(end_vertex) {
            println!("ERROR: 開始頂点 '{}' または 終了頂点 '{}' がグラフに存在しません。", start_vertex, end_vertex);
            return (None, std::u32::MAX);
        }

        // 距離を初期化: 全ての頂点への距離を無限大に設定し、開始頂点のみ0とする
        let mut distances: HashMap<String, u32> = self.data.keys()
            .map(|vertex| (vertex.clone(), std::u32::MAX))
            .collect();
        distances.insert(start_vertex.to_string(), 0);

        // 経路を記録するための辞書: 各頂点に至る直前の頂点を保持
        let mut predecessors: HashMap<String, Option<String>> = self.data.keys()
            .map(|vertex| (vertex.clone(), None))
            .collect();

        // 優先度付きキュー: (距離, 頂点) のタプルを格納し、距離が小さい順に取り出す
        let mut priority_queue = BinaryHeap::new();
        priority_queue.push(State {
            cost: 0,
            vertex: start_vertex.to_string(),
        });

        while let Some(State { cost, vertex }) = priority_queue.pop() {
            // 取り出した頂点への距離が、すでに記録されている距離より大きい場合はスキップ
            if cost > distances[&vertex] {
                continue;
            }

            // 終了頂点に到達したら探索終了
            if vertex == end_vertex {
                break;
            }

            // 現在の頂点から到達可能な隣接頂点を探索
            if let Some(neighbors) = self.get_neighbors(&vertex) {
                for (neighbor, weight) in neighbors {
                    let distance_through_current = match distances[&vertex].checked_add(*weight) {
                        Some(val) => val,
                        None => continue, // オーバーフロー防止
                    };

                    // より短い経路が見つかった場合
                    if distance_through_current < distances[neighbor] {
                        distances.insert(neighbor.clone(), distance_through_current);
                        predecessors.insert(neighbor.clone(), Some(vertex.clone()));
                        // 優先度付きキューに隣接頂点を追加または更新
                        priority_queue.push(State {
                            cost: distance_through_current,
                            vertex: neighbor.clone(),
                        });
                    }
                }
            }
        }

        // 終了頂点への最短距離が無限大のままなら、到達不可能
        if distances[end_vertex] == std::u32::MAX {
            println!("INFO: 開始頂点 '{}' から 終了頂点 '{}' への経路は存在しません。", start_vertex, end_vertex);
            return (None, std::u32::MAX);
        }

        // 最短経路を再構築
        let mut path = Vec::new();
        let mut current = end_vertex.to_string();
        while current != start_vertex {
            path.push(current.clone());
            if let Some(Some(prev)) = predecessors.get(&current) {
                current = prev.clone();
            } else {
                // これは起こらないはずだが、安全のため
                return (None, std::u32::MAX);
            }
        }
        path.push(start_vertex.to_string());
        path.reverse(); // 経路は逆順に構築されたので反転

        (Some(path), distances[end_vertex])
    }
}

// ヒューリスティック関数 (この例では常に0、ダイクストラ法と同じ)
fn dummy_heuristic(_u: &str, _v: &str) -> u32 {
    0
}

fn main() {
    println!("Dijkstra -----> start");

    let mut graph_data = GraphData::new();

    graph_data.clear();
    let input_list = vec![
        ("A", "B", 4), ("B", "C", 3), ("B", "D", 2), 
        ("D", "A", 1), ("A", "C", 2), ("B", "D", 2)
    ];
    for &(v1, v2, weight) in &input_list {
        graph_data.add_edge(v1, v2, weight);
    }
    println!("\nグラフの頂点: {:?}", graph_data.get_vertices());
    println!("グラフの辺 (重み付き): {:?}", graph_data.get_edges());
    let input = ("A", "B");
    let shortest_path = graph_data.get_shortest_path(input.0, input.1, dummy_heuristic);
    println!("経路{}-{} の最短経路は {:?} (重み: {})", 
             input.0, input.1, shortest_path.0, 
             if shortest_path.1 == std::u32::MAX { "無限大".to_string() } else { shortest_path.1.to_string() });

    graph_data.clear();
    let input_list = vec![
        ("A", "B", 4), ("C", "D", 4), ("E", "F", 1), ("F", "G", 1)
    ];
    for &(v1, v2, weight) in &input_list {
        graph_data.add_edge(v1, v2, weight);
    }
    println!("\nグラフの頂点: {:?}", graph_data.get_vertices());
    println!("グラフの辺 (重み付き): {:?}", graph_data.get_edges());
    let input = ("A", "B");
    let shortest_path = graph_data.get_shortest_path(input.0, input.1, dummy_heuristic);
    println!("経路{}-{} の最短経路は {:?} (重み: {})", 
             input.0, input.1, shortest_path.0, 
             if shortest_path.1 == std::u32::MAX { "無限大".to_string() } else { shortest_path.1.to_string() });

    graph_data.clear();
    let input_list = vec![
        ("A", "B", 4), ("B", "C", 3), ("D", "E", 5)
    ];
    for &(v1, v2, weight) in &input_list {
        graph_data.add_edge(v1, v2, weight);
    }
    println!("\nグラフの頂点: {:?}", graph_data.get_vertices());
    println!("グラフの辺 (重み付き): {:?}", graph_data.get_edges());
    let input = ("A", "D");
    let shortest_path = graph_data.get_shortest_path(input.0, input.1, dummy_heuristic);
    println!("経路{}-{} の最短経路は {:?} (重み: {})", 
             input.0, input.1, shortest_path.0, 
             if shortest_path.1 == std::u32::MAX { "無限大".to_string() } else { shortest_path.1.to_string() });

    graph_data.clear();
    let input_list: Vec<(&str, &str, u32)> = vec![];
    for &(v1, v2, weight) in &input_list {
        graph_data.add_edge(v1, v2, weight);
    }
    println!("\nグラフの頂点: {:?}", graph_data.get_vertices());
    println!("グラフの辺 (重み付き): {:?}", graph_data.get_edges());
    let input = ("A", "B");
    let shortest_path = graph_data.get_shortest_path(input.0, input.1, dummy_heuristic);
    println!("経路{}-{} の最短経路は {:?} (重み: {})", 
             input.0, input.1, shortest_path.0, 
             if shortest_path.1 == std::u32::MAX { "無限大".to_string() } else { shortest_path.1.to_string() });

    println!("\nDijkstra <----- end");
}